src/main/java/com/google/opensesame/projects/ProjectQuery.java

```diff
@@ -0,0 +1,127 @@
+package com.google.opensesame.projects;
+
+import static com.googlecode.objectify.ObjectifyService.ofy;
+
+import com.google.opensesame.util.ErrorResponse;
+import com.googlecode.objectify.annotation.Index;
+import com.googlecode.objectify.cmd.Query;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Helper functions for querying the projects Datastore. */
+public class ProjectQuery {
+  public static final String FILTER_QUERY_PARAM = "filter";
+
+  /**
+   * The fields of ProjectEntity that have the @Index Objectify annotation and can be used for
+   * sorting and filtering in Datastore.
+   */
+  public static List<Field> queryableFields;
+
+  static {
+    Field[] fields = ProjectEntity.class.getDeclaredFields();
+    queryableFields = new ArrayList<Field>();
+    for (Field field : fields) {
+      if (field.isAnnotationPresent(Index.class)) {
+        queryableFields.add(field);
+      }
+    }
+  }
+
+  /**
+   * Gets the query filters from a servlet request. Returns null if any errors are encountered while
+   * parsing the query filters. See {@link com.google.opensesame.projects.QueryFilter.fromString}
+   * for more information on how QueryFilters are parsed.
+   *
+   * @param HttpServletRequest The request to get the filter queries from.
+   * @param HttpServletResponse The servlet response to send errors to.
+   * @return Returns the list of QueryFilters or null if an error was encountered while parsing.
+   */
+  private static List<QueryFilter> getQueryFiltersFromRequest(
```

Comment:
Just an idea, not sure why `getQueryFiltersFromRequest` or `queryFromRequest` has to return `null` in any case. Right now the errors are all handled by the ErrorResponse, so returning null is just another special case the callers will have to handle. Why not just raise the exception? Or even just return an empty list or collection.

Reply:
This is definitely a symptom of the separation of concerns issue we identified in a previous PR (#150). This is definitely not ideal and could be described as a "hacky" way to indicate to the caller of the method that an error has occurred and a response has been sent.

The reason this needs to be differentiated from returning an empty list is that an empty list indicates to the caller/client that their query returned no results, versus a query that could not be completed being unable to return results at all.

So I'm thinking the best course of action is to get working on issue #150, move errors to a wrapper class, and avoid having to use this `null` return value in any of these methods.

Reply:
Actually, on the "Why not just raise the exception," do you think it'd be better to refactor this to use a wrapper class similar to the one used [in my portfolio](https://github.com/Richie78321/google-step/blob/master/portfolio/src/main/java/com/google/sps/helper/ValidationResult.java), or do you think raising a custom exception would be a better and more native solution?

Reply:
It seems like raising an exception should be saved for [truly exceptional cases](https://stackoverflow.com/a/77361/9017299), and in this case I think it's best to assume that the string can be invalid, so I'm thinking wrapper class, but would like your opinion when you get the chance.

Reply:
I think we're in agreement, this class shouldn't have to know about `HttpServletRequest` or `HttpServletResponse` objects.

Error wrapper or exception is fine. I was leaning toward throwing an exception. I don't see "truly exceptional cases" in the answer you linked? I do get that reading from the second-most upvoted answer.

The first answer (linked) talks about throwing exceptions when pre-assumptions are broken. Seems like that's exactly what's happening here, if I have a function to parse a query from a string then it seems reasonable to me to expect that string to be in a particular format. And if it isn't, that would be where an exception comes in.

A plus of exceptions is that you get the stack trace for free.

Reply:
Okay, sounds good. I'll refactor this.

Reply:
@billyeh Just finished refactoring. I love the result! Much cleaner in usage and in testing. Thanks so much for the suggestion.